### PR TITLE
Nano ID v2.0

### PR DESCRIPTION
**Pull request by telesvar**
All the changes and tests are added

### CODE REVIEW

**Comment by puyuan** [comment]
I am going to point the PR to v2.0 branch first.  v2.0 branch contains the changes I intend to merge to master.

File: nanoid/complex_generate.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# coding: utf-8
+
+from alphabet import alphabet
```
Comment (puyuan) [inline nanoid/complex_generate.py]: @aidarkhanov 
Please reorder packages according to PEP. 
standard modules
third party modules
your own modules

File: test/nanoid_test.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# coding: utf-8
+
+from nanoid import generate
+from pytest import approx
+
+alphabet = '_-0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
+
+
+def test_generates_url_friendly_id():
```
Comment (puyuan) [inline test/nanoid_test.py]: @aidarkhanov 
I am not sure about the style of this test case.  Should we used the standard python unittest.Testcase style?

File: nanoid/non_secure_generate.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# coding: utf-8
+
+from alphabet import alphabet
```
Comment (puyuan) [inline nanoid/non_secure_generate.py]: @aidarkhanov   Same, the order of the packages

File: test/nanoid_test.py (hunk 2 above)
Comment (telesvar) [reply test/nanoid_test.py]: Sure enough, pytest is easier option. Do not know why tests should migrate to standard unittest.

Comment (puyuan) [reply test/nanoid_test.py]: I am kind of a python standard guy. pytest is a framework, and frameworks change over time. And people could well be running this in IDE or some other test framework.

Comment (telesvar) [reply test/nanoid_test.py]: Your points are all valid, but pytest became a standard in terms of testing. Though, I will try to port them to standard python as the tests are not really complicated enough.

**Comment by puyuan** [comment]
For naming, I am not sure if fast_generate and generate is the characteristics of the ID generation.  At least we haven't proven its  orders of magnitude fast. 

In the js repo, the naming is  generate,  non-secure/generate.  It more agrees with the underlying implementation.  Fast and slow is a performance benchmark and is more fickle, depending on the python library, os and implementations. 

I suggest the following:

1. 
nanoid.generate()
nanoid.non_secure_generate()

or   

2.  nanoid.generate(secure=True)  or False

And if you decide to go with 1. 
   For the  generate function, you can merge the _complex_generate logic directly, no need 2 function.  e.g. one public, one private.  The logic seems simple enough

   And the fast_generate function is fine, just change to   non_secure_generate  ?

**Comment by telesvar** [comment]
> For naming, I am not sure if fast_generate and generate is the characteristics of the ID generation. At least we haven't proven its orders of magnitude fast.
> 
> In the js repo, the naming is generate, non-secure/generate. It more agrees with the underlying implementation. Fast and slow is a performance benchmark and is more fickle, depending on the python library, os and implementations.
> 
> I suggest the following:
> 
> 
> nanoid.generate()
> nanoid.non_secure_generate()
> 
> or
> 
> 1. nanoid.generate(secure=True)  or False
> 
> And if you decide to go with 1.
> For the generate function, you can merge the _complex_generate logic directly, no need 2 function. e.g. one public, one private. The logic seems simple enough
> 
> And the fast_generate function is fine, just change to non_secure_generate ?

Non secure generate is not significantly faster to name it so, I will remove private functions to simplify the library

**Comment by aidarkhanov** [comment]
Seems like everything is done for sure

**Comment by telesvar** [comment]
Seems like everything is done for sure

**Comment by puyuan** [comment]
I will review the changes and see if any remaining conflicts, and do the merge.

**Comment by puyuan** [comment]
@aidarkhanov  While doing the merge and running some of my existing test cases, I noticed a bug. Can you check and help fix it?  With a shorter alphabet, some id length is not generated correctly.

```
    def test_short_secure_ids(self):
        for i in range(10000):
            nanoid = generate(alphabet="12345a", size=3)
>           self.assertEqual(len(nanoid), 3)
E           AssertionError: 2 != 3


```

File: test/nanoid_test.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            if distribution < min:
+                min = distribution
+        self.assertLessEqual(max - min, 0.05)
+
+    def test_has_no_collisions(self):
+        count = 100 * 1000
+        used = {}
+        for _ in range(count):
+            id = generate()
+            self.assertIsNotNone(used.has_key(id))
```
Comment (puyuan) [inline test/nanoid_test.py:49]: @aidarkhanov  I noticed has_key has been removed for python 3 dictionary.

Comment (telesvar) [reply test/nanoid_test.py:49]: Ok, I will fix the issues

**Pull request closed without merging**